quic/s2n-quic-core/src/havoc.rs

```diff
@@ -0,0 +1,630 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    frame::{self, new_connection_id::STATELESS_RESET_TOKEN_LEN},
+    stream::StreamType,
+    varint,
+};
+use core::{convert::TryInto, ops::Range};
+pub use s2n_codec::{Encoder, EncoderBuffer, EncoderValue};
+
+pub trait Random {
+    fn fill(&mut self, bytes: &mut [u8]);
+
+    #[inline]
+    fn shuffle(&mut self, bytes: &mut [u8]) {
+        if bytes.is_empty() {
+            return;
+        }
+
+        let count = self.gen_range(0..bytes.len());
+        for _ in 0..count {
+            let from = self.gen_range(0..bytes.len());
+            let to = self.gen_range(0..bytes.len());
+            bytes.swap(from, to);
+        }
+    }
+
+    #[inline]
+    fn gen_slice<'a>(&mut self, bytes: &'a mut [u8]) -> &'a mut [u8] {
+        if bytes.is_empty() {
+            return bytes;
+        }
+
+        let len = self.gen_range(0..bytes.len());
+        let bytes = &mut bytes[..len];
+        self.fill(bytes);
+        bytes
+    }
+
+    #[inline]
+    fn gen_bool(&mut self) -> bool {
+        self.gen_u8() & 0b1 == 0b1
+    }
+
+    #[inline]
+    fn gen_u8(&mut self) -> u8 {
+        let mut o = [0];
+        self.fill(&mut o);
+        o[0]
+    }
+
+    #[inline]
+    fn gen_u32(&mut self) -> u32 {
+        let mut o = [0; 4];
+        self.fill(&mut o);
+        u32::from_le_bytes(o)
+    }
+
+    #[inline]
+    fn gen_u64(&mut self) -> u64 {
+        let mut o = [0; 8];
+        self.fill(&mut o);
+        u64::from_le_bytes(o)
+    }
+
+    #[inline]
+    fn gen_varint(&mut self) -> varint::VarInt {
+        use varint::VarInt;
+        let max = VarInt::MAX.as_u64().try_into().unwrap_or(usize::MAX);
+        let v = self.gen_range(0..max);
+        unsafe { VarInt::new_unchecked(v as _) }
+    }
+
+    #[inline]
+    fn gen_range(&mut self, range: Range<usize>) -> usize {
+        let value = self.gen_u64() as usize;
+        value.max(range.start).min(range.end.saturating_sub(1))
+    }
```

Comment:
This doesn't seem random? I would expect something more like https://github.com/aws/s2n-quic/blob/main/quic/s2n-quic-core/src/packet/stateless_reset.rs#L130